### PR TITLE
ENH Support more array-like data types for tabular data and list-like data types for text data

### PR DESCRIPTION
**Pull request by anferico**
This pull requests is an attempt towards addressing #65. Note that this is my first ever public contribution, so don't be too hard on me 😉 

For `tabular-classification` and `tabular-regression` tasks, we're currently supporting the following data types:

- pandas.DataFrame
- numpy.ndarray

Should this PR get merged, we will support any other data type that can be converted to a 2D numpy.ndarray using `np.asarray`.

---

For `text-classification` and `text-regression` tasks, we're currently supporting the following data types:

- List of strings

Should this PR get merged, we will support any iterable of strings.

xref: [https://github.com/skops-dev/skops/pull/45/files#r933221139](https://github.com/skops-dev/skops/pull/45/files#r933221139)

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks a lot @anferico for taking this.

I have only taken a quick look so far and noticed that you haven't added any unit tests for the changes. Could you please do that? Let us know if you need any help with that.

Also, please add an entry to the [changes doc](https://github.com/skops-dev/skops/blob/main/docs/changes.rst).

**Comment by anferico** [comment]
I haven't added any unit test for the 2 small helper functions I added, though I've modified existing ones in accordance to the changes I made. Not sure what's your policy here, but if needed I can add unit tests for those as well. I also added some docstrings and modified existing ones.

Everything *should* be fine since all unit tests were passing on my local machine, but let's wait until the pytest workflow runs successfully.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:269]: Same, can we have a test?

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:266]: I'm undecided on this one. Here we're testing all the data, even though we only need the first 3 samples. Of course, we can argue that fitting the model should not work if some rows have strange types, but is it really our job here to verify this?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thank you very much for working on this. To me, this looks already quite good and the overall approach can stay the same. However, I have some small comments, mostly regarding efficiency and some missing tests. Please take a look.

**Comment by anferico** [comment]
@BenjaminBossan I've tried addressing your comments, see if it's good now. One minor detail: I renamed the function `_is_iterable_of_strings` to `_is_sequence_of_strings` because now I'm not expecting generic iterables anymore as inputs to that function, which led me to remove the logic that prevents iterators and generators from being exhausted when iterated over. Without that logic, it just felt wrong to me calling it `_is_iterable_of_strings`, so I went for `_is_sequence_of_strings`.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:137]: Now the data is sliced twice

File: skops/hub_utils/tests/test_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/tests/test_hf_hub.py:283]: I think it would be nice to additionally test:

- a numpy array of strings
- an empty container
- for generators, that they're not consumed (could be a separate test)

File: skops/hub_utils/tests/test_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/tests/test_hf_hub.py:301]: Could you please add an additional test for numpy arrays of strings and for empty containers?

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:249]: This part is getting rather complex and distracts from the actual work being done by `_create_config`. I wonder if it would be possible to factor this code into `_get_example_input`, so that this whole block here could be simplified to:

```python
    config["sklearn"]["example_input"] = _get_example_input(data)
    if "tabular" in task:
        config["sklearn"]["columns"] = _get_column_names(data)
```

If that's not really possible, just leave it as is.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:254]: At some point, we will probably have to factor out functions like `_validate_folder`, `_head`, `_is_sequence_of_strings` etc. to a utils module. It doesn't need to be in this PR though.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for making the changes, the feature is almost there. I still have a few comments left, some are just suggestions that could be tackled later. Please also update with the main branch, as there is a merge commit.

File: skops/hub_utils/_hf_hub.py
Comment (anferico) [reply skops/hub_utils/_hf_hub.py:249]: How about something like this?
```python
config["sklearn"]["example_input"] = _get_example_input(data, task)
if "tabular" in task:
    config["sklearn"]["columns"] = _get_column_names(data)
```
Cause this would make it easier for me, I think. Alternatively, I'm thinking of something like:
```python
if "tabular" in task:
    config["sklearn"]["example_input"] = _get_example_input_from_tabular_data(data)
    config["sklearn"]["columns"] = _get_column_names(data)
elif "text" in task:
    config["sklearn"]["example_input"] = _get_example_input_from_text_data(data)
```
although the latter would imply renaming an existing function

Comment (BenjaminBossan) [reply skops/hub_utils/_hf_hub.py:249]: > config["sklearn"]["example_input"] = _get_example_input(data, task)

Looks good to me.

File: docs/changes.rst
Comment (adrinjalali) [inline docs/changes.rst:14]: line too long (limit to 79 chars)

File: skops/hub_utils/_hf_hub.py
Comment (adrinjalali) [inline skops/hub_utils/_hf_hub.py:77]: we should use `sklearn.utils.validation.check_array` instead of writing our own, unless we handle more input types.

File: skops/hub_utils/_hf_hub.py
Comment (adrinjalali) [inline skops/hub_utils/_hf_hub.py:116]: we shouldn't just convert to numpy, for instance, if the input is xarray, we can extract column names. The same with cuDF.

File: skops/hub_utils/_hf_hub.py
Comment (adrinjalali) [inline skops/hub_utils/_hf_hub.py:281]: to me it seems better to remove this method and have it in the function where it's called.

File: skops/hub_utils/_hf_hub.py
Comment (adrinjalali) [inline skops/hub_utils/_hf_hub.py:254]: same here, not sure if we need a method for this.

**Review by adrinjalali — CHANGES_REQUESTED** [review]
Thanks for the PR @anferico

File: skops/hub_utils/_hf_hub.py
Comment (anferico) [reply skops/hub_utils/_hf_hub.py:116]: Do you have a specific list of data types that you'd like to support? Cause there are countless variants of numpy arrays and pandas dataframes, but I don't think we can find a general solution that works for each of them

Comment (adrinjalali) [reply skops/hub_utils/_hf_hub.py:116]: I don't have a list, but I'm not sure if a silent conversion is a good idea. It might be a better idea to fail on containers which we don't understand, and slowly add them as requests come, WDYT?

Comment (anferico) [reply skops/hub_utils/_hf_hub.py:116]: Ok, so if I understood correctly you'd rather fail on, say, cuDF dataframes rather than silently converting them to NumPy arrays, and the reason is because you don't want to risk throwing away information about the column names. Is that right?

Comment (adrinjalali) [reply skops/hub_utils/_hf_hub.py:116]: Yes. And the reason I don't want to miss that information is the following:

We use these feature names to figure out how to pass the information to the model in the inference side. And that API receives data in this format: `{'colA': [list of values], ...}`. If we lose the feature name information here, the user wouldn't know how to send data to the inference backend and the whole inference side of things break. Since saving feature names here serves pretty much no other purpose than making the inference backend, it'd be a pity to silently convert them to numpy arrays and lose that information.

So I think it's better to fail on things we don't understand, and explicitly add the more popular ones. WDYT?

Comment (anferico) [reply skops/hub_utils/_hf_hub.py:116]: Makes total sense to me, however this means we need to explicitly check the type of the inputs we get, right? So, basically, right now we're supporting only:

- `DataFrame`s (for which we can also extract column names)
- `np.array`s
- [list|tuple] of [list|tuple|np.array] (can be silently converted to `np.array`s using sklearn's `check_array`)

Anything else we should add for now?

Comment (adrinjalali) [reply skops/hub_utils/_hf_hub.py:116]: I think this would be a good start. I'm curious to see what people are gonna use. We should ideally raise for all other types or at least raise a warning.

**Comment by BenjaminBossan** [comment]
@anferico are you still working on this?

**Comment by anferico** [comment]
> @anferico are you still working on this?

I am, it's just that I haven't had much free time recently. I'm happy to continue working on this, but if it's really high priority please feel free to let someone else take over. Sorry about that

**Comment by BenjaminBossan** [comment]
Hey, no worries, take the time you need. I just asked because sometimes people just forget or are waiting on some input from the maintainers.

**Comment by anferico** [comment]
@BenjaminBossan I think I've addressed all the points above, however now I'm facing another problem. 
I noticed that when calling the `test_get_column_names_pandas_not_installed` function, I incur into a weird recursion error. Obviously, as a result, the test fails.
Here's the command I ran:
```bash
python -m pytest -s -v --cov-report=xml skops/hub_utils/tests/test_hf_hub.py::test_get_column_names_pandas_not_installed
```
and here is part of the output I get (see  [log_test.txt](https://github.com/skops-dev/skops/files/10184639/log_test.txt) for the full output):
```
  .
  .
  .
  File "/home/anfri/MyProjects/skops/skops/conftest.py", line 13, in mock_import
  File "/home/anfri/miniconda3/envs/skops/lib/python3.10/unittest/mock.py", line 1103, in __call__
  File "/home/anfri/miniconda3/envs/skops/lib/python3.10/unittest/mock.py", line 1111, in _increment_mock_call
  File "/home/anfri/MyProjects/skops/skops/conftest.py", line 13, in mock_import
  File "/home/anfri/miniconda3/envs/skops/lib/python3.10/unittest/mock.py", line 1103, in __call__
  File "/home/anfri/miniconda3/envs/skops/lib/python3.10/unittest/mock.py", line 1111, in _increment_mock_call
RecursionError: maximum recursion depth exceeded
```
I'm not sure what's happening here, because I haven't touched `conftest.py`. My suspicion is that the line `from sklearn.utils import check_array` that I added to `_hf_hub.py` could be the culprit because there are some pandas imports in there, so maybe us patching the `import` builtin in such a way that it throws an `ImportError` while trying to import pandas could be a problem, but I'm really not sure.

**Comment by BenjaminBossan** [comment]
> I'm not sure what's happening here, because I haven't touched `conftest.py`. My suspicion is that the line `from sklearn.utils import check_array` that I added to `_hf_hub.py` could be the culprit because there are some pandas imports in the source file containing that function, so maybe us patching the `import` builtin in such a way that it throws an `ImportError` while trying to import pandas could be a problem, but I'm really not sure.

It seems likely that this is the explanation, but I'm not sure why that would lead to an infinite recursion. I'd have to dig deeper. @adrinjalali any ideas?

Meanwhile, since some time has passed since this PR was opened, the `changes.rst` has been updated. You would need to move your entry to the current  version. Could you please do that?

**Comment by adrinjalali** [comment]
No idea about the infinite recursion, but looking at this, two things come to my mind:

- it might be a good idea for us to have a CI job where we have minimal dependencies installed, instead of patching load_module.
- same issues as #211 would apply here: we should check if what we do would work on the inference API side.

It might be a good idea to reduce the scope of this PR to list of strings only, to make it easier to merge and tackle other things in a separate PR.

**Comment by BenjaminBossan** [comment]
@anferico I figured out the recursion error. Honestly, I'm a bit surprised that it didn't occur previously, but patching can be tricky. Please change the `pandas_not_installed` fixture like so:

```python
@pytest.fixture
def pandas_not_installed():
    # patch import so that it raises an ImportError when trying to import
    # pandas. This works because pandas is only imported lazily.
    orig_import = __import__

    def mock_import(name, *args, **kwargs):
        if name == "pandas":
            raise ImportError
        return orig_import(name, *args, **kwargs)

    with patch("builtins.__import__", side_effect=mock_import):
        yield
```

**Comment by BenjaminBossan** [comment]
> * it might be a good idea for us to have a CI job where we have minimal dependencies installed, instead of patching load_module.

You mean a CI job where, for instance, pandas is not installed? Yes, that could be useful, but I still like to have a test that patches `import`, so that I can e.g. quickly test the functionality locally without the need to juggle multiple environments.

The extra CI environment would come in handy in case that the test itself is buggy or incomplete. However, having such a test environment would probably require adjusting a bunch of other tests too that currently assume all `tests` libraries to be installed.

> * same issues as [Added support for Structured Arrays #211](https://github.com/skops-dev/skops/pull/211) would apply here: we should check if what we do would work on the inference API side.

Would it be possible to add this in a separate PR?

**Comment by adrinjalali** [comment]
> Would it be possible to add this in a separate PR?

what does "this" here refer to?

**Comment by BenjaminBossan** [comment]
> what does "this" here refer to?

I mean checking if what we do would work on the inference API side.

**Comment by adrinjalali** [comment]
> I mean checking if what we do would work on the inference API side.

Yes and no 😁 we should at least check if we train a model with these data types, scikit-learn understands them (in terms of feature names etc), and that we can predict using these types and don't get errors/warnings from sklearn.

Right now sklearn supports feature names mostly through pandas only.

**Comment by anferico** [comment]
@BenjaminBossan thanks, the recursion error just got fixed. I also fixed the `changes.rst` file.

A couple of comments regarding my latest commit:

- I removed support for generators as valid containers for text-based tasks. The reason is that I couldn't find a way to peek at a generator without consuming it. Let me know if this is ok
- The function `get_example_input_from_text_data` does not fail on empty containers. Is it desirable or not?

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:137]: I don't think it's necessary to have an extra one-line local function defined for this purpose. This creates overhead when it's not necessary and can make the code harder to read, because I cannot simply read from top to bottom but have to jump around.

My guess is that you do this for readability, so that you can later do:

```python
if is_subscriptable(data):
    return data[:n]
```

instead of

```python
if hasattr(data, "__getitem__"):
    return data[:n]
```

which can be a bit more difficult to understand. However, the same effect can be achieved by assigning an extra variable:

```python
is_subscriptable = hasattr(data, "__getitem__")
if is_subscriptable:
    return data[:n]
```

Personally, I would also be happy with the 2nd case, but 3rd one is fine.

For the other locally defined functions, I think a similar argument can be made, though it's not quite as clear cut. I leave it up to you.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:148]: There isn't really a need to assign the error message here. It is better to move it into the `except` part below.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/_hf_hub.py:260]:
```suggestion
```

File: skops/hub_utils/tests/test_hf_hub.py
Comment (BenjaminBossan) [inline skops/hub_utils/tests/test_hf_hub.py:555]: I think this test could also be parametrized. Furthermore, could you please add a test for tuples? I think they are much more likely to be used than sets. Also, how about one example with only 1 element?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Overall, I'm quite happy with the PR. I think it has developed quite well and the code should be easy to maintain.

I have a couple of comments which should hopefully be easy to address, please take a look.

> I removed support for generators as valid containers for text-based tasks. The reason is that I couldn't find a way to peek at a generator without consuming it. Let me know if this is ok

Yes, it's okay. We can add them later if we want to. It is possible to peek at the head of a generator without consuming it completely, but of course the first few elements would need to be loaded at that point. One way to get there is by using [`tee`](https://docs.python.org/3/library/itertools.html#itertools.tee).

> The function `get_example_input_from_text_data` does not fail on empty containers. Is it desirable or not?

Good question. I think it's fine for it not to raise an error. Maybe we can warn, not sure. @skops-dev/maintainers  WDYT?

File: skops/hub_utils/_hf_hub.py
Comment (anferico) [reply skops/hub_utils/_hf_hub.py:137]: Makes sense, thanks for the feedback. If you don't mind, I would change your code slightly so that the sentence has also a subject, namely:
```python
is_data_subscriptable = hasattr(data, "__getitem__")
if is_data_subscriptable:
    return data[:n]
```

File: skops/hub_utils/_hf_hub.py
Comment (anferico) [reply skops/hub_utils/_hf_hub.py:148]: The reason I assigned the error message there is because otherwise I would have to duplicate the same string in the `else` and `except` branches, like so:
```python
try:
    data_head = _head(data, n=3)
    if _is_sequence_of_strings(data_head):
        return {"data": data_head}
    else:
        raise ValueError("The data needs to be a sequence of strings.")
except TypeError as e:
    raise ValueError("The data needs to be a sequence of strings.") from e
```
Another way to avoid this would be to raise a `TypeError` inside the `else` branch, although that would feel pretty hacky.
Do you think it's fine to duplicate the error message?

Comment (BenjaminBossan) [reply skops/hub_utils/_hf_hub.py:148]: Sorry, I have missed that it's used at two places.

File: skops/hub_utils/_hf_hub.py
Comment (BenjaminBossan) [reply skops/hub_utils/_hf_hub.py:137]: That's perfectly fine.

**Comment by anferico** [comment]
@BenjaminBossan  I've addressed the latest comments, let me know if this can be merged.

File: docs/changes.rst
Comment (BenjaminBossan) [inline docs/changes.rst:29]: I didn't notice earlier, but since we released v0.4 in December, this should be added to a new section "v0.5".

**Review by BenjaminBossan — APPROVED** [review]
Apart from the one comment below, this LGTM. Before merging, I would just wait to see if @adrinjalali also wants to review or is fine with going ahead.

**Review by adrinjalali — APPROVED** [review]
Nice! Other than the changelog issue, LGTM.

**Pull request merged**
